src/evaluations/evaluator.py

```diff
@@ -224,38 +237,94 @@ def __init__(self, evaluation_config, sketch_estimator_config_list, run_name,
+    # Spawnable process to evaluate a sketch estimator on a certain scenario.
+    def run_scenario(args):
+      """
+      Tests each scenario on the given estimator.
+
+      Returns:
+        tuple: (elapsed time, estimator name, scenario name)
+      """
+      scenario_config, sketch_estimator_config = args
+      # logging.info(f'Estimator: {sketch_estimator_config.name}'
+      #         f'=== Scenario: {scenario_config.name}')
+      start_time = time.perf_counter()
+      # This is where the bulk of the work happens
+      self.run_one_scenario(scenario_config, sketch_estimator_config)
+      elapsed_time = time.perf_counter() - start_time
+      # logging.info('Finished')
+
+      return elapsed_time, sketch_estimator_config.name, scenario_config.name
+
+    # Get all combinations of (scenario, estimator)
+    work_items = itertools.product(
+                  self.evaluation_config.scenario_config_list,
+                  self.sketch_estimator_config_list
+                 )
+    total_length = len(self.evaluation_config.scenario_config_list) * \
+                len(self.sketch_estimator_config_list)
+
+    # Progress visualization
+    pbar = tqdm(total=total_length)
+
+    # Start scenarios, returns the elapsed time for each.
+    with ProcessPool(self.parallel_cores) as pool:
+      times = pool.uimap(run_scenario, work_items)
+    
+    # While the scenarios are running, save estimator configs
     for sketch_estimator_config in self.sketch_estimator_config_list:
-      start_time = time.time()
-      self.evaluate_estimator(sketch_estimator_config)
-      elapsed_time = str(time.time() - start_time)
+      self.save_estimator(sketch_estimator_config)
+
+    # Aggregate time spent on each scenario by file 
+    # (which aggregates them by estimator)
+    performance_stats = dict()
+
+    # The 'times' uimap will return items which finish first
+    # We are summing up the time each evaluator spends on all
+    # threads. This should provide similar runtime numbers as
+    # the serial version because it calculates CPU time, not
+    # wall time.
+    for elapsed_time, sketch_estimator_name, scenario_name in times:
       time_file = os.path.join(
-          self.description_to_file_dir[KEY_ESTIMATOR_DIRS][
-              sketch_estimator_config.name],
-          EVALUATION_RUN_TIME_FILE)
+        self.description_to_file_dir[KEY_ESTIMATOR_DIRS][
+            sketch_estimator_name],
+        EVALUATION_RUN_TIME_FILE)
+      if time_file not in performance_stats.keys():
+        performance_stats[time_file] = 0
+      performance_stats[time_file] += elapsed_time
+      # Update progress
+      pbar.update()
+      pbar.set_description(f'Finished {sketch_estimator_name}-{scenario_name}')
+
+    # Write performance stats
+    for time_file, elapsed_time in performance_stats.items():
       with open(time_file, 'w') as f:
-        f.write(elapsed_time)
+        f.write(str(elapsed_time))
+    pbar.close()
```

Comment:
Is it possible to use pbar with a context manager?

Reply:
Yes, would that be preferable? I tend to avoid using with statements if they wrap something relatively large

Reply:
With respect to wrapping something large, make some helper functions. The loop that starts on line 271 is a good candidate for turning into a private method called "wait_for_worker_completion".  With that out of the way the amount of code the with statement wraps will be much simpler.

Using context managers to ensure cleanup is a really good practice. Here maybe it doesn't matter that much, in the sense that I'm not sure if tqdm acquires resources. Maybe if you don't close it the pbar your terminal ends up needing to be reset? 

In retrospect though, I'm not sure taking the dependency on tqdm is worth it.

@jgoodknight WDYT?

Reply:
Resolved in 2bcd89f

Reply:
Tqdm is a pretty cool addition with strongly positive usability benefits to folks running long jobs.  I'm a big fan of it and it's super common in the ml world; it's a dependency in TensorFlow if I'm not mistaken.